### ISSUE TITLE
Pretty URL handling for CMS content

### ISSUE THREAD

**Issue by timrobertson100** [gbif/portal16#23]
Rather than having e.g. ```/article/123``` it is often the case that editors designate a ```pretty``` URL for a page (e.g. ```/licenses```).

This needs to be reflected in the web app, meaning that the routing of URLs needs to be driven by content sourced from the CMS.  The CMS needs to surface this in the API and the web app needs to consider these in it's routing.

**Comment by burkeker** [comment]
As the latest Content API will provide for content typed news, individual node will have "self" property like `/api/v1.0/news/newsroom/news/bid-2015-africa-call-for-full-proposals` which uses URL alias, while it's still possible to get the same piece by `/api/v1.0/news/[nid]`. The same will apply for each item in search results. `news` in the url is used for indicating the content type, which is also available in both node JSON and search result so the URL can be constructed when needed in other scenarios.

**Comment by timrobertson100** [comment]
Is that a typo:
`/api/v1.0/news/newsroom/news/bid-2015-africa-call-for-full-proposals`

`.../news/newsroom/news/...` seems very redundant.

**Comment by burkeker** [comment]
It's not a typo. So two ways of accessing the node:
`/api/v1.0/news/[URL alias]` and `/api/v1.0/news/[nid]`. URL alias is assigned by editors and we have /newsroom/news/bid-2015-africa-call-for-full-proposals at the moment for that piece. From the Express' point of view it doesn't need to know because it'll be available from the search result already.

**Comment by timrobertson100** [comment]
So one way that could work:

- User types http://www.gbif.org/newsroom/news/bid-2015-africa-call-for-full-proposals which hits the news controller in express
- Given we will have a better URL structure in the new site, we will probably redirect the user to http://www.gbif.org/news/bid-2015-africa-call-for-full-proposals or similar
- Controller issues an http://api.gbif.org/v1/news/bid-2015-africa-call-for-full-proposals to the backend which is Drupal 
- Drupal recognises this as an alias and does either returns a `HTTP 301` to e.g. http://api.gbif.org/v1/news/123, or returns the JSON and adds a `Link: <http://api.gbif.org/v1/news/123>; rel="canonical"` to indicate that it is an aliased response to the canonical URL.

I don't think it wise to try and historical URL aliases in the design of a new content API. `/api/v1.0/news/newsroom/news/bid-2015-africa-call-for-full-proposals` is a pretty obscure URL to support.

Alternative approaches to the above could be 
  1. explicitly having a redirect service in the CMS that can call to find the correct API to call - e.g. `/locate?url=/newsroom/news/bid-2015-africa-call-for-full-proposals` which would return ´http://api.gbif.org/v1/news/123´
  2. having a service which surfaces all aliases and their target URLs in 1 call, that express caches and uses as a lookup, refreshing that in the background periodically on e.g. 1 minute timer

**Comment by burkeker** [comment]
Agree it's a cleaner URL structure, though that'll require necessary clean-up and redirections to setup. We can plan that.

**Comment by timrobertson100** [comment]
Following discussion here we opted for the following.

- All CMS managed content can optionally have a `targetURL` property which is the preferred URL for the page.  Internally, this is implemented as the `alias` field in the Drupal content model.  Any changes to the alias will result in Drupal storing internally a `redirect` from the old alias to the new alias.
- Any visitor to the GBIF.org website using an old, current, or identifier based URL (e.g. /news/1) will always receive a `301` to the latest alias (e.g. www.gbif.org/news/1 would redirect you to www.gbif.org/news/bid-calls-open) should an alias exist.

To achieve this, the CMS API needs to provide a lookup service, and a GET by the identifier.  

***An example of the calls to the CMS API backend illustrates this***


1. User comes to GBIF using www.gbif.org/newsroom/news/bid-call-for-proposals, a very old url

    ```
GET /lookup?q=/newsroom/news/bid-call-for-proposals
{
  id: 1,
  "type": "news",
  "targetUrl": "/news/bid-calls"  
}
```
2. The webapp would then 301 redirect the user to www.gbif.org/news/bid-calls which is the new, preferred URL.  When the webapp receives this incoming redirection, it looks up again

    ```
GET /news/bid-calls
{
  id: 1,
  "type": "news",
  "targetUrl": "/news/bid-calls"  
}
```
3. The webapp app then recognises the user is on the correct page, and can construct the backend call

    ```
GET /news/1
{
  id: 1,
  "type": "news"
  "targetUrl": "/news/bid-calls",
  "title": "Calls for proposal in the BID project are now open!"
  "description": ...
}
```

Note: The same is true if a user visits www.gbif.org/news/1.  The webapp would notice that the `targetURL` is not the URL the user visited, and would redirect the user to the www.gbif.org/news/bid-calls page and they would then start at 2) above.

**Comment by burkeker** [comment]
The Express layer now handles the URL alias and redirection according to the lookup from the CMS API side:
http://devnode-vh.gbif.org:3000/news/first-bid-grants-for-africa => with targetURL
http://devnode-vh.gbif.org:3000/news/newsroom/news/first-bid-grants-for-africa => with old URL alias
http://devnode-vh.gbif.org:3000/news/first-bid-grantssss-for-africa => URL alias that doesn't exist

**Comment by burkeker** [comment]
BTW the link of the search result to articles has also been updated to targetUrl.

**Comment by timrobertson100** [comment]
I think that should work should it not?

http://devnode-vh.gbif.org:3000/newsroom/news/first-bid-grants-for-africa

**Comment by burkeker** [comment]
You're right. Fixing.

**Comment by burkeker** [comment]
Fixed as ff8b46488888f9a2458fd3a319fcbef5b02d717f.
